### PR TITLE
Node clearcoat normals (2)

### PR DESCRIPTION
**Pull request by sunag**
PR - https://github.com/mrdoob/three.js/pull/17240

@arobertson0 I started a new PR from some principles of your PR. 

The design of `nodes/accessor` is the values not change in flow of generating code, for this reason, your `geometryNormal` is good to replace the old `normal`. 

`normal` and `clearCoatNormal` originally is a result of variation of  `geometryNormal`, for this reason,  `NormalNode.NORMAL` use now `geometryNormal`, but that will be edited in the future to `NormalNode.VIEW`.

In other things I managed to simplify the PR, reusing the nodes already created.
As the changes seemed big, I decided to create a new PR to make it easier.

image

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js [hunk 1]
```diff
@@ -20,14 +20,10 @@ geometry.position = - vViewPosition;
...
 geometry.viewDir = normalize( vViewPosition );
 
-#ifdef USE_CLEARCOAT_NORMALMAP
+#ifdef PHYSICAL
 
 	geometry.clearCoatNormal = clearCoatNormal;
 
-#else
-
-	geometry.clearCoatNormal = geometryNormal;
```
Comment (mrdoob) [inline src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Hmm, was this change intentional?

Comment (sunag) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Yes, `clearCoatNormal` `variable` is every declared if the material is `PHYSICAL`, now this is independent of has defined a `material.clearCoatNormalMap`. This too fix the design proposed in `lights_fragment_begin` of not having any native MAPS, only native LIGHTS.

Comment (mrdoob) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Hmm, I'll have to study this closely tomorrow but. I added this check so `material.normalMap` doesn't affect clear coat when `material.clearCoatNormalMap` is not defined.

That's how the top-right material works here:
https://raw.githack.com/mrdoob/three.js/dev/examples/webgl_materials_clearcoat_normalmap.html

Does the change still support hat behavior?

Comment (mrdoob) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Hmm, looking at the code a bit closely... yeah, I think it does.

Comment (mrdoob) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: I guess we can also do this:

```glsl
	#ifdef PHYSICAL
	vec3 clearCoatNormal;
	#endif
```

https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/common.glsl.js#L42

Comment (sunag) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Yeah,  `geometryNormal = normal` is seted before of `NormalMap`, in this sense `geometryNormal` is a pure normal. 

After `clearCoatNormal = geometryNormal` is seted. 

`normal` can be a variation of `material.normalMap` and `clearCoatNormal` a variation of `material.clearCoatNormalMap`. 

`normalMap` and `clearCoatNormalMap` both dont not depend on each other. Both in theory is derived from `geometryNormal`.

Comment (sunag) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: Added, I too made this change `ifnref STANDARD` to `ifdef PHYSICAL`, I think that this is more readable https://github.com/mrdoob/three.js/pull/17256/commits/b244ee366dc865019946d2be246d9d2d6f871534

Comment (sunag) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: With relation of `NormalNode.NORMAL`, now this use `geometryNormal`, in this sense, `NormalNode.NORMAL` is not affected with changes of `normal` or `clearCoatNormal`.

Comment (arobertson0) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: > Added, I too made this change `ifndef STANDARD` to `ifdef PHYSICAL`, I think that this is more readable [b244ee3](https://github.com/mrdoob/three.js/commit/b244ee366dc865019946d2be246d9d2d6f871534)

I think this is great, but other defines would have to be changed.

I've been confused by this myself. It seems like in some places, the code assumes `PHYSICAL` is the inverse of `STANDARD` (e.g.: _lights_fragment_maps.glsl.js_), while in other places it's assume `STANDARD` implies `PHYSICAL` (e.g.: _envmap_pars_fragment.glsl.js_).

@DanielSturk and I have been talking about making the defines more expressive. This mix-up would be covered in that PR. What are your opinions?

Comment (sunag) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: I think, make this system more modular and it can help performance in customized materials like NodeMaterial for example, instead of having only two materials, a full (PHYSICAL) and another half (STANDARD) add intermediates modules in `STANDARD` that are activated in `PHYSICAL` declaration, for example:

```javascript
#ifdef PHYSICAL
	#define CLEAR_COAT
	#define SHEEN
	#define TRANSPARENCY
#endif
```

Comment (arobertson0) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js]: > I think, make this system more modular and it can help performance in customized materials like NodeMaterial

I agree, this is what we were looking at too.

**Comment by mrdoob** [comment]
Thanks!

**Comment by arobertson0** [comment]
It look like the clearcoat normals don't work with the envmap. This is because `ReflectNode` reflects off the geometry normals and not the clearcoat normals. 

With cleracoat normals, we are breaking the assumption that there is only one normal surface. This means we have to be able to specify the normal for `ReflectNode`. I know it would mean removing the _unique_ tag, but I think this is how we get the environment to reflect  correctly.

I've had to do the same in _envmap_physical_pars_fragment.glsl.js_. `reflectVec` is now a function of which normal we are reflecting from.

**Comment by arobertson0** [comment]
Edit: The environment doesn't reflect off the base layer's normals either. It's reflecting smoothly everywhere. You can see this by commenting out `scene.add( lightGroup );` on the material nodes example.

**Comment by sunag** [comment]
Hmm I see ... I think in this case we can contextualize the `normal`, from `Node.getUnique` we can use `unique` tag in a variable way. @arobertson0 Thanks 👍
